File: nuido_flow_data/static/src/components/data/reference_map_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { useState } from "@odoo/owl";
import { Node } from "@nuido/components/node";
import { ModelSelectorEx } from "@nuido_flow/components/ui/model_selector_ex";
import { ModelFieldSelectorEx } from "@nuido_flow/components/ui/model_field_selector_ex";
import { TextInputDialogInput } from "@nuido_base/components/dialogs/text_input_dialog_input";
export class ReferenceMapNode extends Node {
    setup() {
        super.setup();
        this.state = useState({
            model: this.props.node.model,
            modelDescription: this.props.node.model_description,
            fieldName: "",
            fieldValue: ""
        });
    }
    onModelSelected(model) {
        const { label, technical: value } = model;
        this.props.node.fields = [];
        this.state.model = value;
        this.state.modelDescription = label;
        this.props.node.model = value;
        this.props.node.model_description = label;
    }
    async onModelFieldUpdate(path, info) {
        this.props.node.field = info.fieldDef ? info.fieldDef.name : "";
        this.refreshEdges();
    }
    onReferenceChanged(value) {
        this.props.node.reference = value;
    }
    get modelSelectorId() {
        return `input-${this.props.node.id}}-model-selector`;
    }
    get fieldValueId() {
        return `input-${this.props.node.id}}-field-value`;
    }
}
ReferenceMapNode.template = "nuido_flow_data.reference-map-node";
ReferenceMapNode.components = {
    ...Node.components,
    ModelSelectorEx,
    ModelFieldSelectorEx,
    TextInputDialogInput
};

File: nuido_flow_data/flows/data/reference_map_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.

import logging

from odoo.tools import safe_eval

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode
from odoo.addons.nuido_flow.flows.tools.tools import get_default_context_for_eval, get_active_record_info

from odoo.addons.nuido_flow.flows.tools.log_const import LOGGER_NAME
_logger = logging.getLogger(LOGGER_NAME)

class ReferenceMapNode(BaseNode):
    def _process(self, params):
        super()._process(params)

        context = get_default_context_for_eval(self.env)
        if params is not None:
            context['params'] = params

        info = get_active_record_info(self.env)
        context = {**context, **info}

        reference = self.definition["reference"]

        try:
            reference_value = safe_eval.safe_eval(reference, context)
        except:
            _logger.warning("Exception on evaluation: %s", reference, exc_info=True)
            reference_value = -1

        model = self.definition["model"]
        field = self.definition["field"]
        domain = [(field, "=", reference_value)]

        record = self.env[model].search_read(domain, ["id"])
        if len(record) == 0:
            res = -1
        else:
            res = record[0]

        return {
            "result": res
        }
